### PR TITLE
[QA-325] Adding queue method

### PR DESCRIPTION
**Pull request by Aleksandras-V**

### CODE REVIEW

File: ipc_receivefile.c [hunk 1]
```diff
@@ -418,16 +426,81 @@ int ipc_receive_pipe(char* fileName){
...
+	do {
+		errno = 0;
+		msg_queue = mq_open( "queue", O_RDONLY, &attrs );
+		if (errno == ENOENT){
+			sleep (2);
+		}
+	}while (errno == ENOENT);
+
+	if (msg_queue == -1){
+		perror ("mq_open");
```
Comment (RomainChaumontet) [inline ipc_receivefile.c:454]: free(buffer)

File: ipc_receivefile.c [hunk 2]
```diff
@@ -418,16 +426,81 @@ int ipc_receive_pipe(char* fileName){
...
+
+	do {
+
+		receive_size = mq_timedreceive (msg_queue, buffer, MAX_MSG_QUEUE_SIZE, &prio,&abs_timeout);
+
+		if (receive_size == -1) {
+			if (errno == ETIMEDOUT) {
+				printf ("mq_timedreceive() timed out file fully transfered.\n");
+			 } else {
+				perror ("mq_timedreceive()");
```
Comment (RomainChaumontet) [inline ipc_receivefile.c:470]: free(buffer) 

You should check every exit/return and also in ipc_sendfile

File: ipc_sendfile.c [hunk 3]
```diff
@@ -340,19 +348,80 @@ int ipc_send_pipe(char* fileName){
...
+	printf("Creating Queue... \n");
+
+	memset(&attrs, 0, sizeof attrs);
+	attrs.mq_maxmsg = MAX_MSG_QUEUE;
+	attrs.mq_msgsize = MAX_MSG_QUEUE_SIZE;
+	msg_queue = mq_open( "queue", O_WRONLY | O_CREAT, S_IRWXU | S_IRWXG, &attrs );
+	if (msg_queue == -1) {
+		perror ("mq_open()");
+	    return EXIT_FAILURE;
+	}
```
Comment (RomainChaumontet) [inline ipc_sendfile.c:371]: if the queue already exist and have some message on it, you won't see it and the receivefile will get them first -> the copy won't be exact.
You should unlink the queue first or use O_EXCL to know if the queue already exists.

**Review by RomainChaumontet — CHANGES_REQUESTED** [review]
If the file use to write the copy doesn't exist, it doesn't work.
image
You can just open the file in "w" mode to create if it doesn't exist.

I think you should regroup ipc_sendfile.h and ipc_receivefile.h in a common file to be sure the element on there are the same for both of programs. You can write the name of the queue in this file too.

File: ipc_receivefile.c (hunk 1 above)
Comment (Aleksandras-V) [reply ipc_receivefile.c:454]: Done!

File: ipc_receivefile.c (hunk 2 above)
Comment (Aleksandras-V) [reply ipc_receivefile.c:470]: Done!

File: ipc_sendfile.c (hunk 3 above)
Comment (Aleksandras-V) [reply ipc_sendfile.c:371]: Fixed!

**Review by RomainChaumontet — APPROVED** [review]

**Comment by pra-mah** [comment]
While receiving message , I can only give file which already exists.

**Pull request merged**